Fix water flooding onto lava

juhdanad opened the pull request:
Lava cooled by water now clears its light.
Fixes the bug introduced by #4655.
Sorry about that bug...

(EDIT by paramat, author's comment below copied up to here)

A little help to reviewers: `MapBlock::getLightingExpired()` and `MapBlock::isGenerated()` always returned the same value, so I removed the former.
Instead I introduced `MapBlock::getLightingIncomplete()`, which returns 16 bits (`u16`). Each bit indicates if in a direction in a particular light bank light propagation/removal was not finished, because the neighbor block was unloaded.
When the server loads a block, it checks these flags and corrects lighting if necessary.

To introduce the new flags I had to increase the map block serialization version, this is why you can not play on your maps on other versions of Minetest after you opened them in this version.
Of course old clients still can play, just the server has to be new to be able to read the map.

paramat commented:
Will test.

paramat commented:
How do you reproduce this bug? Does it need basic shaders on?

juhdanad commented:
The bug is independent of the shaders. You will see it with and without shaders.
Just place a lava block on a mountain, then let it flow down, finally place water over the lava's source.
@Fixer-007 found the bug: https://github.com/minetest/minetest/pull/4655#issuecomment-256614945

Fixer-007 commented:
@juhdanad Wait, you mean ghost lighting after lava casts? If so, that bug was for ages before you, sorry for not clarifying that :( It is awesome if you actually fixed them, you will be in my hall of fame of fixes.

paramat commented:
Yes that's an old bug.

Fixer-007 commented:
Ok. I did some testing with huge lava casts, bright ghost lighting is fixed.

juhdanad commented:
@Fixer-007, how did you create stone over water?

sofar commented:
He makes lava flow over water, not water over lava.

Fixer-007 commented:
@juhdanad thats lava casts, you can make them by placing some water on big slopes and then higher you place water sources, after that you wait until it completes, then you can remove lava sources.

Also, I seen you asked someone to help about TNT light bugs? I can help you. Contact me on #minetest-dev

juhdanad commented:
These ghost lights seem to appear at the border of loaded and unloaded area. My code assumes that the light is correct everywhere except the currently changed nodes, but this is not always true: if lava disappears near to an unloaded block, its light still will be in the unloaded block.
You can see that on the third picture by @Fixer-007: the border of the light is sharp.
I can imagine two solutions to this:
1. emerge areas when updating lighting.
2. when loading a block, check if its light is valid (it is enough to test the borders). Update the light if necessary.
I would prefer the second solution.
What do you think? Do you have other ideas to have correct light even in unloaded blocks?

Fixer-007 commented:
Will check!

juhdanad commented:
There is a Travis compiler error with unordered maps of `v3s16` and `MapBlock*`, do you have any suggestions?

juhdanad commented:
Sorry, I forgot to mention: direct sunlight correction is not included to support floatlands.

Fixer-007 commented:
Will try to compile in mingw64

paramat commented:
Remove WIP label?

paramat commented:
> direct sunlight correction is not included to support floatlands.

Mgv7 floatlands or typical floating rocks in mgv5 or mgv7?

Fixer-007 commented:
Mingw64 compilation success.

juhdanad commented:
I was writing about Mgv7's floatlands.
They have 'featured lighting bugs' as their shadows terminate in mid-air.
This code itself can not resume propagation of these shadows.
But if I implemented the correction of direct sunlight and the block where a floatland's shadow terminates were marked as having 'unfinished lighting', the shadows could propagate further downwards.

So if sunlight were corrected and I (or somebody else) modified light calculation at map generation to support the 'unfinished lighting' flag, the floatlands' shadows would appear on the ground level, too.

paramat commented:
Ok good. Obviously the floatland shadows should continue to terminate.

juhdanad commented:
In my opinion the best solution would be to add 'giant fluorescent fungi' (or other sunlight-emitting node) at the bottom of floatlands to make sunlight bug-free, after the feature freeze.

Fixer-007 commented:
It won't fix already present ghost lighting, right?

juhdanad commented:
Yes, only new bugs won't appear.
This implementation is performance-efficient, therefore does not search for lighting bugs if not necessary.

Fixer-007 commented:
Testing right now: I wonder if those guys are different bug: https://i.imgur.com/HNbAlEE.png

Fixer-007 commented:
No ghost lighting after first huge attempt: https://i.imgur.com/Ew2Cfao.png

juhdanad commented:
I have a strong feeling that this is a mesh generation bug (if the node had light, it would be uniformly lit). If you test with smooth lighing disabled, you won't see such bugs.

Fixer-007 commented:
No ghost lighting on second attempt. 
BTW, I've tried to reproduce different shadow bug: big square shadow after sporadic TNT explosions and succeeded, but shadow was few levels down only.

paramat commented:
> In my opinion the best solution would be to add 'giant fluorescent fungi' (or other sunlight-emitting node) at the bottom of floatlands to make sunlight bug-free

I guess you mean to fix the sky-going-black bug. Adding glowing nodes under floatlands occured to me too, if this actually works.

juhdanad commented:
@paramat yes, please remove WIP label.

Fixer-007 commented:
I have hard time reproducing the ghost lighting, I think it is fixed.

paramat commented:
Seems to have an issue on build.

juhdanad commented:
Yes, unordered map of v3s16 and Mapblock* causes the compilation error, but I don't know why. (Something is requested there)

juhdanad commented:
Should I use plain std::map?

juhdanad commented:
Problems fixed!

paramat commented:
Although a fix would be nice for release, it is complex changes to lighting so i feel there is potential for creating new bugs that don't get caught before release, which is close. This also lacks reviews.
Perhaps best left until after release.

Fixer-007 commented:
Meanwhile I'm compiling with it by default, and have not encountered anything suspicious, when it will be reviewed it will be already tested for some time by at least me.

juhdanad commented:
Thanks, @Fixer-007!

Fixer-007 commented:
Have not had any problems with it to this day, I guess after release developers should review it and merge. Only thing is you won't be able to play on maps created with never minetest unfortunately on older versions of minetest (because of change in mapblock code - it displays the warning about unsupported format because of new mapblock revision).

juhdanad commented:
Rebased! I think it is well tested thanks to @Fixer-007 and now waiting for reviews!

juhdanad commented:
A little help to reviewers: `MapBlock::getLightingExpired()` and `MapBlock::isGenerated()` always returned the same value, so I removed the former.
Instead I introduced `MapBlock::getLightingIncomplete()`, which returns 16 bits (`u16`). Each bit indicates if in a direction in a particular light bank light propagation/removal was not finished, because the neighbor block was unloaded.
When the server loads a block, it checks these flags and corrects lighting if necessary.

To introduce the new flags I had to increase the map block serialization version, this is why you can not play on your maps on other versions of Minetest after you opened them in this version.
Of course old clients still can play, just the server has to be new to be able to read the map.

paramat commented:
Perhaps you could convince hmmmm (kwolekr on ithub) to review your light PRs, i think he should see them especially due to the changes to vm:update_map etc. and to voxelmanip stuff.

Fixer-007 commented:
I'm confused, is not this PRs commit is also included in https://github.com/minetest/minetest/issues/4967 ?

paramat commented:
It is, but only because that PR is based on this.

Zeno- commented on src/map.cpp:
```diff
@@ -3005,71 +2987,89 @@ MapBlock* ServerMap::loadBlock(v3s16 blockpos)
+			}
+			catch(InvalidFilenameException &e)
+			{
+				return NULL;
```
If you're catching these shouldn't they be logged? (otherwise why not just catch them all and return NULL?)

juhdanad replied:
In fact I just indented these lines.

Zeno- commented on src/mapblock.h:
```diff
@@ -213,17 +213,42 @@ class MapBlock /*: public NodeContainer*/
+	inline u16 getLightingIncomplete()
 	{
-		return m_lighting_expired;
+		return m_lighting_incomplete;
```
I'm not 100% sure I like this -- it's almost like a double negative. Personally I'd prefer this function was called isLightingComplete() and the member called m_lighting_complete

Zeno- commented on src/mapblock.h:
```diff
@@ -213,17 +213,42 @@ class MapBlock /*: public NodeContainer*/
+		setLightingIncomplete(newflags);
+	}
+
+	inline bool isLightingIncomplete(LightBank bank, u8 direction)
```
See above

Zeno- commented on src/voxelalgorithms.cpp:
```diff
@@ -747,6 +765,125 @@ void update_lighting_nodes(Map *map, INodeDefManager *ndef,
+			brightest_neighbor = light2;
+		}
+	}
+	return brightest_neighbor == light + 1;
```
I haven't really looked, but is it at all possible that light + 1 can overflow?

juhdanad replied:
No, light ≤ 15.

paramat replied:
But 15 + 1 would overflow?

Zeno- replied:
add assert then so it's obvious and we don't have to go searching.

Zeno- commented on src/voxelalgorithms.cpp:
```diff
@@ -747,6 +765,125 @@ void update_lighting_nodes(Map *map, INodeDefManager *ndef,
+		return true;
+	}
+	u8 light = n.getLightNoChecks(bank, &f);
+	u8 brightest_neighbor = f.light_source + 1;
```
I haven't really looked, but is it at all possible that f.light_source + 1 can overflow?

juhdanad replied:
light_source is always ≤15, it is checked at `ContentFeatures:deSerialize()`. I will add an assert, though.

paramat replied:
You mean node light source is always <= 14?

juhdanad replied:
Yes, ≤14.

paramat commented:
> I was writing about Mgv7's floatlands.
They have 'featured lighting bugs' as their shadows terminate in mid-air.
This code itself can not resume propagation of these shadows.
But if I implemented the correction of direct sunlight and the block where a floatland's shadow terminates were marked as having 'unfinished lighting', the shadows could propagate further downwards.
So if sunlight were corrected and I (or somebody else) modified light calculation at map generation to support the 'unfinished lighting' flag, the floatlands' shadows would appear on the ground level, too.

This might not be a problem because any light updates near the shadow cutoff will already propagate the shadow downwards, for example if someone builds at the shadow cutoff and builds downwards to ground level the shadow will reach ground level.
So if you prefer you could perhaps implemement direct sunlight correction if it's a good idea.

paramat commented:
Or maybe you mean there is a danger at mapgen time?

juhdanad commented:
@paramat there are two flags in m_lighting_complete that I would like to use in sunlight correction. But I would prefer that in a separate PR, as I plan to add sunlight-emitting nodes first. Without them, sunlight correction would cause huge shadows under floatlands (as soon as they are generated).

paramat commented:
Ok i see, yes a separate PR is best.

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -747,6 +765,127 @@ void update_lighting_nodes(Map *map, INodeDefManager *ndef,
+					MapNode n = b->getNodeNoCheck(x, y, z,
+						&is_valid_position);
+					u8 light = n.getLight(bank, ndef);
+					// Sunlight is fix
```
Is this meant to be "Sunlight is fixed"?

juhdanad replied:
We have done direct sunlight spreading before, and it is correct, no update needed.
So sunlight is fix/constant from now on.

paramat replied:
"Sunlight is fixed" is correct English.

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -747,6 +765,127 @@ void update_lighting_nodes(Map *map, INodeDefManager *ndef,
+		for (u8 i = 0; i <= LIGHT_SUN; i++) {
+			const std::vector<ChangingLight> &lights = light_sources.lights[i];
+			for (std::vector<ChangingLight>::const_iterator it = lights.begin();
+				it < lights.end(); it++) {
```
One more indent?

paramat commented:
Looked through the code. Although i don't have a complete understanding of all of it i have no objections. I would like to +1 but am hesitant too, maybe after looking through again i will. I would like other devs to review this as they have better C++ skills than me.

paramat commented on src/map.cpp:
```diff
@@ -3005,71 +2987,89 @@ MapBlock* ServerMap::loadBlock(v3s16 blockpos)
+		int loadlayout = 1;
+		std::string sectordir1 = getSectorDir(p2d, 1);
+		std::string sectordir;
+		if(fs::PathExists(sectordir1))
```
Space after 'if'

paramat commented on src/map.cpp:
```diff
@@ -3005,71 +2987,89 @@ MapBlock* ServerMap::loadBlock(v3s16 blockpos)
-			return NULL;
+			loadlayout = 2;
+			sectordir = getSectorDir(p2d, 2);
 		}
```
The conditional above should be formatted:
```
		if (fs::PathExists(sectordir1)) {
			sectordir = sectordir1;
		} else {
			loadlayout = 2;
			sectordir = getSectorDir(p2d, 2);
		}
```

paramat commented on src/map.cpp:
```diff
@@ -3005,71 +2987,89 @@ MapBlock* ServerMap::loadBlock(v3s16 blockpos)
+			{
+				return NULL;
+			}
 		}
```
```
		if (sector == NULL) {
			try {
				sector = loadSectorMeta(sectordir, loadlayout != 2);
			} catch(InvalidFilenameException &e) {
				return NULL;
			} catch(FileNotGoodException &e) {
				return NULL;
			} catch(std::exception &e) {
				return NULL;
			}
		}
```

paramat commented on src/map.cpp:
```diff
@@ -3005,71 +2987,89 @@ MapBlock* ServerMap::loadBlock(v3s16 blockpos)
-	if(fs::PathExists(sectordir + DIR_DELIM + blockfilename) == false)
-		return NULL;
+		std::string blockfilename = getBlockFilename(blockpos);
+		if(fs::PathExists(sectordir + DIR_DELIM + blockfilename) == false)
```
Space after 'if'

paramat commented on src/map.cpp:
```diff
@@ -3005,71 +2987,89 @@ MapBlock* ServerMap::loadBlock(v3s16 blockpos)
+		std::map<v3s16, MapBlock*> modified_blocks;
+		// Fix lighting if necessary
+		voxalgo::update_block_border_lighting(this, block, modified_blocks);
+		if(!modified_blocks.empty()) {
```
Space after 'if'

paramat commented on src/map.cpp:
```diff
@@ -3005,71 +2987,89 @@ MapBlock* ServerMap::loadBlock(v3s16 blockpos)
+			std::map<v3s16, MapBlock *>::iterator it;
+			for (it = modified_blocks.begin();
+					it != modified_blocks.end(); ++it)
+			event.modified_blocks.insert(it->first);
```
Needs indent.
I assume this line is the single line looped in the 'for' statement.

paramat commented on src/mapblock.h:
```diff
@@ -213,17 +213,42 @@ class MapBlock /*: public NodeContainer*/
-		if (expired != m_lighting_expired){
-			m_lighting_expired = expired;
-			raiseModified(MOD_STATE_WRITE_NEEDED, MOD_REASON_SET_LIGHTING_EXPIRED);
+		if (newflags != m_lighting_complete){
```
Space after '()'

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -747,6 +765,127 @@ void update_lighting_nodes(Map *map, INodeDefManager *ndef,
+				continue;
+			}
+			// Only update if lighting was not completed.
+			if(block->isLightingComplete(bank, d) &&
```
Space after 'if'

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -747,6 +765,127 @@ void update_lighting_nodes(Map *map, INodeDefManager *ndef,
+			}
+			// Only update if lighting was not completed.
+			if(block->isLightingComplete(bank, d) &&
+				other->isLightingComplete(bank, 5 - d)) {
```
2 indents.

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -747,6 +765,127 @@ void update_lighting_nodes(Map *map, INodeDefManager *ndef,
+			}
+			// Reset flags
+			block->setLightingComplete(bank,d, true);
+			other->setLightingComplete(bank,5-d, true);
```
Space after ',' and around '-'.

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -747,6 +765,127 @@ void update_lighting_nodes(Map *map, INodeDefManager *ndef,
+				continue;
+			}
+			// Reset flags
+			block->setLightingComplete(bank,d, true);
```
Space after ','

paramat commented:
Added line comments.

juhdanad commented:
All requested changes are done!

paramat commented:
Ok i think that as long as another dev with more C++ knowledge than me reviews this i can give a second +1.

paramat commented:
Important point that needs to be considered:
> To introduce the new flags I had to increase the map block serialization version, this is why you can not play on your maps on other versions of Minetest after you opened them in this version.
Of course old clients still can play, just the server has to be new to be able to read the map.

paramat commented:
In another few days i'll go over this again and can hopefully then +!.

paramat commented:
>  this is why you can not play on your maps on other versions of Minetest after you opened them in this version.

This may be controversial, i will gather opinions.

paramat commented:
Fixer-007 wrote
> thats lava casts, you can make them by placing some water sources on big slopes and then higher you place lava sources, after that you wait until it completes, then you can remove lava sources and inspect for ghost light. Sometimes you need few tries to catch the lights.

Since this bug is not common, is it worth the change in mapblock serialisation? I'm not sure. I have added a post in the forum to gather opinions https://forum.minetest.net/viewtopic.php?f=18&t=16595

juhdanad commented:
I think I could re-write the other pull request not to depend on this, but that would make that less efficient. TNTs exploding next to unloaded area could still cause lighting bugs then.

paramat commented:
Ok. seeing as version change seems essential for the 2nd PR i think it's worth it then.
I can see how the 1st PR will be useful for much more than just huge lavacooling mountains.
So, i'll go through the 1st one again then will probably +1.
Then i'll start looking at the 2nd.

paramat commented on src/mapblock.h:
```diff
@@ -105,7 +105,7 @@ class NodeContainer
 #define MOD_REASON_REALLOCATE                (1 << 1)
 #define MOD_REASON_SET_IS_UNDERGROUND        (1 << 2)
-#define MOD_REASON_SET_LIGHTING_EXPIRED      (1 << 3)
+#define MOD_REASON_SET_LIGHTING_INCOMPLETE   (1 << 3)
```
Should this be `MOD_REASON_SET_LIGHTING_COMPLETE` ?

paramat replied:
@juhdanad

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -596,6 +600,22 @@ void update_lighting_nodes(Map *map, INodeDefManager *ndef,
 		UnlightQueue disappearing_lights(256);
 		ReLightQueue light_sources(256);
+		// Nodes that are brighter than the brightest modified node was
+		// won't change, since they didn't got their light from a
```
got -> get

paramat commented on src/voxelalgorithms.cpp:
```diff
@@ -596,6 +600,22 @@ void update_lighting_nodes(Map *map, INodeDefManager *ndef,
+			}
+		}
+		// If only one node changed, even nodes with the same brightness
+		// didn't got their light from the changed node.
```
got -> get

paramat commented:
Added line comments, minor issues.
I have a basic understanding of this now so a cautious :+1: but another dev with more C++ skill than me should review this.

paramat commented:
@Zeno- are you able to review / approve?

juhdanad commented:
@paramat Done!

paramat commented:
:+1:

sofar commented:
:+1:

Ekdohibs commented:
Hmm, why not initialize complete lighting for previously-existing blocks before this PR to incomplete, so that it is recomputed?

paramat commented:
@juhdanad could you squash this down to a minimum number of commits?

juhdanad commented:
@Ekdohibs for performance reasons, this code only corrects lighting bugs if they appear on the border of a block. Setting lighting to incomplete would not automatically correct all of the old lighting bugs.

paramat commented:
Ok for me as it is.

juhdanad commented:
You should wait for Travis IMO (to see if there are any warnings).

paramat commented:
Will merge in 1-2hrs.

paramat commented:
f17c9c45dc30a388675d46418d278a4a029206e2
